x/superfluid/keeper/migrate.go

```diff
@@ -60,41 +74,96 @@ func (k Keeper) UnlockAndMigrate(ctx sdk.Context, sender sdk.AccAddress, lockId
-		return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, fmt.Errorf("shares to migrate must be less than or equal to shares in lock")
+		return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, 0, fmt.Errorf("shares to migrate must be less than or equal to shares in lock")
 	}
 
 	// Exit the balancer pool position.
 	exitCoins, err := k.gk.ExitPool(ctx, sender, poolIdLeaving, sharesToMigrate.Amount, sdk.NewCoins())
 	if err != nil {
-		return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, err
+		return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, 0, err
 	}
 	// Defense in depth, ensuring we are returning exactly two coins.
 	if len(exitCoins) != 2 {
-		return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, fmt.Errorf("Balancer pool must have exactly two tokens")
+		return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, 0, fmt.Errorf("Balancer pool must have exactly two tokens")
 	}
 
 	// Create a full range (min to max tick) concentrated liquidity position.
-	positionId, amount0, amount1, liquidity, joinTime, err = k.clk.CreateFullRangePosition(ctx, concentratedPool, sender, exitCoins)
+	// This will also create a lock for the time remaining from the gamm lock and begin unlocking from where it left off.
+	//
+	// TODO: If we eventually want to support migrating a SFS position from balancer to CL and continue staking,
+	// we need to make a createPosition that does not start unlocking automatically. Right now, this is just being used
+	// as a pseudo unfreeze logic.
+	positionId, amount0, amount1, liquidity, joinTime, concentratedLockId, err = k.clk.CreateFullRangePositionUnlocking(ctx, concentratedPool, sender, exitCoins, remainingLockTime)
 	if err != nil {
-		return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, err
+		return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, 0, err
+	}
+
+	// Link the newly created concentratedLockId to the intermediate account if is currently superfluid delegated
+
+	// TODO: I feel like this is wrong, we need to either send the funds from the old intermediary account to the new one
+	// or somehow ensure the slashing is done on the old intermediary account
+	if isCurrentlySuperfluidDelegated {
+		// Create and set a new intermediary account based on the previous validator but with the new lock id and concentratedLockupDenom
+		concentratedLockupDenom := cl.GetConcentratedLockupDenom(poolIdEntering, positionId)
+		newIntermediaryAcct := types.NewSuperfluidIntermediaryAccount(concentratedLockupDenom, intermediateAccount.ValAddr, concentratedLockId)
+		k.SetIntermediaryAccount(ctx, newIntermediaryAcct)
+
+		// Create a new synthetic lockup for the new intermediary account in an unlocking status
+		err := k.createSyntheticLockup(ctx, concentratedLockId, newIntermediaryAcct, unlockingStatus)
+		if err != nil {
+			return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, 0, err
+		}
+	}
```

Comment:
This TODO is an older comment, but I think a lot of focus needs to be put here still to ensure this is sound along with the code below this block in this file